refactor(semantic): implement `Send` and `Sync` for `ScopingCell`

overlookmotel opened the pull request:
Previous PRs removed unsound `Sync` impl for `Allocator` (#13033) and `Send` impl for `Vec` (#13041).

Previously `ScopingCell` was `Send` and `Sync` (and therefore `Scoping` was too). The recent PRs mean that `ScopingCell` lost those traits too.

This PR implements both traits again for `ScopingCell` by restricting its API slightly, so now it is `Send` and `Sync` again, but this time in a manner which maintains soundness.

The comments in the code outline the logic of why I believe it to be sound.

Rolldown relies on `ScopingCell` being `Send` and `Sync`. After this PR, this stack does not require any changes in Rolldown. I've checked that `cargo ck` in Rolldown passes when using the version of Oxc from this branch.

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/13042?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#13043** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/13043?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#13042** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/13042?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/13042?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#13041** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/13041?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#13040** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/13040?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#13039** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/13039?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#13033** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/13033?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#13032** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/13032?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=13042).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Dunqing reviewed (approved):
Great!

aapoalas commented on crates/oxc_semantic/src/scoping.rs:
```diff
@@ -99,13 +99,123 @@ impl Default for Scoping {
+
+    /// Wrapper around [`ScopingCellInner`], which only provides methods that give access to an
+    /// `&Allocator` reference if provided with `&mut ScopingCell`. See comments above.
+    pub struct ScopingCell(ScopingCellInner);
```
```suggestion
    #[repr(transparent)]
    pub struct ScopingCell(ScopingCellInner);
```

overlookmotel replied:
Is there a reason why this *has* to be `#[repr(transparent)]` (e.g. for soundness)? Or are you suggesting this more as general good practice?

aapoalas replied:
Just general good practice. Since this is intended to be a new-type wrapper over `ScopingCellInner`, we should request that the compiler does so and doesn't by some miracle of magic decide that, eg. while a `ScopingCellInner` is okay to pass in a register, a `ScopingCell` has to be passed on-stack because "it's a struct mate!".

overlookmotel replied:
Aapo shared this link which goes into this in more detail: https://www.reddit.com/r/rust/comments/kbyb6z/comment/gflqiep/

Thanks. I was not aware of this at all. There are many other places we should probably be using `#[repr(transparent)]` then e.g.:

https://github.com/oxc-project/oxc/blob/522356201b712f9eb68740362ea2135dbbb7f9c7/crates/oxc_syntax/src/scope.rs#L17

(currently these types are `#[repr(C)]` due to `#[ast]` macro)

overlookmotel committed:
refactor(semantic): implement `Send` and `Sync` for `ScopingCell` (#13042)

Previous PRs removed unsound `Sync` impl for `Allocator` (#13033) and `Send` impl for `Vec` (#13041).

Previously `ScopingCell` was `Send` and `Sync` (and therefore `Scoping` was too). The recent PRs mean that `ScopingCell` lost those traits too.

This PR implements both traits again for `ScopingCell` by restricting its API slightly, so now it is `Send` and `Sync` again, but this time in a manner which maintains soundness.

The comments in the code outline the logic of why I believe it to be sound.

Rolldown relies on `ScopingCell` being `Send` and `Sync`. After this PR, this stack does not require any changes in Rolldown. I've checked that `cargo ck` in Rolldown passes when using the version of Oxc from this branch.
